guardian/iframe-messenger
# Always round body height up instead of sometimes down

## What does this change?

As noted [on MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetHeight), offsetHeight returns a value that is rounded to the nearest integer. If the height is non-integer it can be rounded down, making the returned height slightly smaller than the actual height of the element.

We recently noticed this is a problem for datawrapper embeds (produced via [datawrapper-import](https://github.com/guardian/datawrapper-import)), e.g. https://interactive.guim.co.uk/datawrapper/embed/XffdG/3/, which has the descender of the g in its source text slightly truncated when embedded into an article.

That MDN page suggests using getBoundingClientRect() instead, which is what this PR does.

getBoundingClientRect() is slightly newer, but based on [the MDN compatibility table](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect#browser_compatibility), I think we should be ok.

Reading through [the link in the readme about element heights](https://docs.google.com/spreadsheets/d/1dv4bO6F9A0KGF6FdLsIJrKnidnrVsM8i3uI-OfCYjWQ/edit?usp=sharing), I'm reassured that this change should work as intended: the only difference between the lines for `document.body.offsetHeight` and `document.body.getBoundingClientRect().height` is that the former is rounded to the nearest integer.

However, maybe instead of reporting a rounded-up size to the parent, we should report the correct (non-integer) size and make the parent round it up? Conceptually that feels more correct (as the size reported by the child would be more correct), even if the behaviour would be the same.

## How to test

- host a new version of the script somewhere with the changes from this PR
- modify datawrapper-import-CODE to link to that new script
- embed a test datawrapper graphic in an article in CODE that would have overflow with the current script
  - this is any graphic with a non-integer pixel height that would round down if rounded to the nearest integer (a height that ends with `.n…` where 1 ≤ n ≤ 4). 
  - for example, [this graphic](https://app.datawrapper.de/archive/team/LjLrk4B8#/jCxow) which when embedded in [this article](https://preview.code.dev-gutools.co.uk/money/2024/nov/18/ghg) has a height of 458.1px.
- see if the overflow is gone

## How can we measure success?

Graphs created by datawrapper-import should no longer have scrollable overflow.

## Have we considered potential risks?

This script is used by a large collection of existing graphics, and if any of them have a height as described above, then this change could make their containers up to 0.4px taller. I think this should be fine (as it should only prevent existing overflow), but it's hard to be certain. It would be good to test with a selection of existing graphics.

## Images



- TODO
